src/loop/parser.ml

```diff
@@ -190,70 +192,131 @@ module Make(State : State.S) = struct
+      (* Formats Dimacs and Tptp are descriptive and lack the emission
+          of formal solve/prove instructions, so we need to add them. *)
+      let s' =
+        match lang with
+        | Logic.Zf
+        | Logic.ICNF
+        | Logic.Smtlib2 _
+        | Logic.Alt_ergo -> s
+        | Logic.Dimacs
+        | Logic.Tptp _ ->
+          Dolmen.Std.Statement.pack [s; Dolmen.Std.Statement.prove ()]
+      in
+      let file = { file with lang = Some lang; } in
+      st, file, (Gen.singleton s')
+
+  (* This is a ['a Gen.t] with a threaded state. *)
+  type 'a gen' = {g: State.t -> State.t * 'a option} [@@unboxed]
+
+  let make g = {g}
+
+  let run {g} = g
+
+  let flat_map (f : 'a -> 'b gen') (g : 'a gen') : 'b gen' =
+    (* [cur_f] holds the last generator returned by a call to [f] *)
+    let cur_f = ref None in
+    let rec aux st =
+      match !cur_f with
+      | None ->
+        (* [cur_f] is empty, get a new one and try again *)
+        let st, x = run g st in
+        begin match x with
+        | None -> st, None
+        | Some x ->
+          cur_f := Some (f x);
+          aux st
+        end
+      | Some c ->
+        (* We got a generator, use it or refill it *)
+        let st, c' = run c st in
+        match c' with
+        | Some _ -> st, c'
+        | None ->
+          cur_f := None;
+          aux st
     in
-    let st = set_logic_file st file in
-    (* Wrap the resulting parser *)
-    st, wrap_parser ~file (Gen.append (Gen.of_list prelude) g)
+    make aux
+
+  (* [map_st_gen f g] calls [f] on all the values produced by [g] in order,
+     threading the state.
+
+     [f] is *not* a generator but a function in the state monad. *)
+  let map_st_gen f (g : _ Gen.t) : _ gen' =
+    make @@ fun st ->
+    begin match g () with
+    | None -> st, None
+    | Some x -> f x st
+    end
+
+  let parse_logic ?(preludes = []) file =
+    Gen.(
+      append
+        (of_list preludes |> map (fun p -> (p, true)))
+        (singleton (file, false)))
+    |> map_st_gen (fun ((file : _ file), is_prelude) ->
+      match file.source with
+      | `Stdin when is_prelude ->
+        fun st -> State.error ~file st stdin_prelude (), None
+      | `Stdin ->
+        fun st ->
+        let st, file, g = parse_stdin st file in
+        let st = State.set State.logic_file file st in
+        st, Some (make @@ wrap_parser ~file g)
+      | `Raw _ | `File _ as source ->
+        (* NB: We need to make sure the lang is set before calling
+            [switch_to_full_mode_if_needed] *)
+        match infer_lang source file with
+        | file, lang ->
+          fun st ->
+          let st, file = switch_to_full_mode_if_needed st file file in
+          let st, file, g = parse_file st source file lang in
+          let st = State.set State.logic_file file st in
+          st, Some (make @@ wrap_parser ~file g)
+        | exception Logic.Extension_not_found ext ->
+          fun st -> State.error st extension_not_found ext, None
```

Comment:
I don't see any reason to split this function as is currently done: we would have the same results if the whole function took both arguments at the begining (and the perf would most certainly be better too), unless I'm missing something.

Reply:
The way this function is split is simply a byproduct of the iterative refactoring that was performed, there is no other reason. I will make that change

Reply:
Actually the whole function can't take both arguments because the split is to eliminate `` `Stdin`` from the variant type, but `infer_lang` is only used here now so I will inline it.